src/Util/WebRootMover.php

```diff
@@ -0,0 +1,323 @@
+                'Version %s is currently installed.',
+                SilverstripePackageInfo::RECIPE_CORE,
+                $packageInfo[0]['version']
+            ));
+        }
+
+        // Make sure we don't already have a public.
+        $publicPath = $this->publicPath();
+        if (file_exists($publicPath) &&
+            (
+                !is_dir($publicPath) ||
+                count(scandir($publicPath)) > 2
+            )
+        ) {
+            throw new InvalidArgumentException(
+                'There\'s already a non empty `public` folder in your project root. ' .
+                'You might alrady be using the public web root.'
+            );
+        }
+    }
+
+    /**
+     * Move server configuration file (.htaccess, web.config) around.
+     * @internal Depends on `checkPrerequisites` having been run.
+     * @param CodeChangeSet $diff
+     */
+    public function moveServerConfigFile(CodeChangeSet &$diff)
+    {
+        $this->processLegacyServerConfigFile($diff, '.htaccess', self::HTACCESS_HASHES);
+        $this->processLegacyServerConfigFile($diff, 'web.config', self::WEBCONFIG_HASHES);
+    }
+
+    /**
+     * Process the provided legacy server config file. The file is expected to be in the root of the project.
+     * If the file is unchanged from previous versions or does not exists, we just override remove it and copy the
+     * recipe core version instead.
+     *
+     * If the file exists and has been changed, we move it to the public folder with a warning.
+     *
+     * @param CodeChangeSet $diff
+     * @param string $filename
+     * @param string[] $unchangedHashes List of valid MD5 hashes used to determine if the file has changed.
+     * @return void
+     */
+    private function processLegacyServerConfigFile(
+        CodeChangeSet &$diff,
+        string $filename,
+        array $unchangedHashes
+    ): void {
+        // We are tracking 2 metrics:
+        // * state of the root server file ... absent, unchanged, changed (3 different scenarios);
+        // * whatever recipe core provides a root file and a public file, or just a public file (2 different scenarios).
+        // So there's 6 different combinations.
+
+
+        // Get reference to the core files.
+        $targetPath = self::PUBLIC . DIRECTORY_SEPARATOR . $filename;
+        $coreRootFile = $this->getRecipeCoreFileContent($filename);
+        $corePublicFile = $this->getRecipeCoreFileContent($targetPath);
+
+        // Try to get the disk item
+        $originFile = $this->disk->exists($filename) ? $this->disk->itemByPath($filename) : false;
+
+        // Figuring out base conditions
+        $originCondition = $originFile ?
+            (in_array($originFile->getMd5Hash(true), $unchangedHashes) ?
+                'unchanged':
+                'changed'
+            ) :
+            'absent';
+        $coreCondition = $coreRootFile ? 'root_file' : 'no_root_file';
+
```

Comment:
I think amount `root_file`, `no_root_file`, `absent`, `changed` and `unchanged` are used, it would make sense to store their values in constants.

Reply:
I wrestled with that question as well. I originally had a bunch of nested if-else statement, but my little brain was hurting looking at it, so I decided to start by checking my condition and then have a flat if-elseif structure.

Wasn't entirely sure how appropriate it was to have class constants to deal the internal logic of a single method. I'll switch to using class constants with a comment that they are purely internal.

Reply:
Fixed.